File: src/bin/day02.rs
use anyhow::Result;
use std::fs::File;
use std::io::{BufRead, BufReader};

fn calculate(data: &[i64]) -> bool {
    let mut increasing = true;
    let mut decreasing = true;
    let mut not_too_large = true;

    for (a, b) in data.iter().zip(data.iter().skip(1)) {
        let diff = b - a;
        if increasing && diff <= 0 {
            increasing = false;
        }

        if decreasing && diff >= 0 {
            decreasing = false;
        }

        if not_too_large && diff.abs() > 3 {
            not_too_large = false;
        }
    }

    (increasing && not_too_large) || (decreasing && not_too_large)
}

struct Permutations<'a> {
    data: &'a [i64],
    skip: usize,
}

impl<'a> Iterator for Permutations<'a> {
    type Item = Vec<i64>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.skip >= self.data.len() {
            return None;
        }

        let mut data = Vec::with_capacity(self.data.len() - 1);

        for (index, item) in self.data.iter().enumerate() {
            if index == self.skip {
                continue;
            }
            data.push(*item);
        }

        self.skip += 1;

        Some(data)
    }
}

fn main() -> Result<()> {
    let buffer = BufReader::new(File::open("data/day02.txt")?);

    let mut part1 = 0;
    let mut part2 = 0;
    for line in buffer.lines() {
        let line = line?;
        let data: Vec<i64> = line.split(' ').map(|x| x.parse().unwrap()).collect();
        if calculate(&data) {
            part1 += 1;
        }
        let mut iter = Permutations {
            data: &data,
            skip: 0,
        };
        while let Some(data) = iter.next() {
            if calculate(&data) {
                part2 += 1;
                break;
            }
        }
    }

    assert_eq!(part1, 591);
    assert_eq!(part2, 621);

    Ok(())
}

File: days/day02.py
data = load_data("day02")

grid = [[int(x) for x in line.split(" ")] for line in data.splitlines()]

# print(grid)


def calculate(line):
    diffs = [b - a for a, b in zip(line[:-1], line[1:])]
    if [x for x in diffs if x >= 0] == [] or [x for x in diffs if x <= 0] == []:
        if len([x for x in diffs if abs(x) in range(4)]) == len(diffs):
            return True
    return False


def permutations(line):
    out = []
    for i in range(len(line)):
        new = line[:]
        new.pop(i)
        out.append(new)
    return out


count = 0
for line in grid:
    if calculate(line):
        count += 1

# check(2, count)
check(591, count)


count = 0
for line in grid:
    if any([calculate(x) for x in permutations(line)]):
        count += 1

# check(4, count)
check(621, count)
